File: integ/postgres.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import pg from "pg";
import {
  CommandSafeguard,
  MountMode,
  PostgresResource,
  Workspace,
} from "@struktoai/mirage-node";

const DSN =
  process.env.POSTGRES_DSN ?? "postgres://mirage:mirage@localhost:55432/mirage_integ";
const MOUNT = "/pg";

const DEC = new TextDecoder();

const BOOKS: ReadonlyArray<readonly [number, string, string, number, number]> = [
  [1, "alpha", "ada", 2020, 4.5],
  [2, "beta", "ben", 2021, 3.2],
  [3, "gamma", "cara", 2022, 5.0],
  [4, "delta", "ada", 2023, 4.0],
  [5, "epsilon", "ben", 2024, 2.5],
];

const AUTHORS: ReadonlyArray<readonly [number, string, number]> = [
  [1, "ada", 2],
  [2, "ben", 2],
  [3, "cara", 1],
];

const CASES: ReadonlyArray<readonly [string, string]> = [
  ["ls_root", `ls ${MOUNT}/`],
  ["ls_schema", `ls ${MOUNT}/public/`],
  ["ls_tables", `ls ${MOUNT}/public/tables/`],
  ["ls_views", `ls ${MOUNT}/public/views/`],
  ["ls_entity", `ls ${MOUNT}/public/tables/books/`],
  ["tree", `tree -L 3 ${MOUNT}/public/`],
  ["cat_schema_json", `cat ${MOUNT}/public/tables/books/schema.json`],
  ["cat_rows", `cat ${MOUNT}/public/tables/books/rows.jsonl`],
  ["head_2", `head -n 2 ${MOUNT}/public/tables/books/rows.jsonl`],
  ["tail_2", `tail -n 2 ${MOUNT}/public/tables/books/rows.jsonl`],
  ["wc_l_books", `wc -l ${MOUNT}/public/tables/books/rows.jsonl`],
  ["wc_default_books", `wc ${MOUNT}/public/tables/books/rows.jsonl`],
  ["wc_l_authors", `wc -l ${MOUNT}/public/tables/authors/rows.jsonl`],
  ["stat_rows", `stat ${MOUNT}/public/tables/books/rows.jsonl`],
  ["grep_c_title", `grep -c title ${MOUNT}/public/tables/books/rows.jsonl`],
  ["grep_ada", `grep ada ${MOUNT}/public/tables/books/rows.jsonl`],
  ["grep_schema_scope", `grep ada ${MOUNT}/public/tables/`],
  ["rg_schema_scope", `rg ben ${MOUNT}/public/`],
  ["find_rows", `find ${MOUNT}/public/ -name rows.jsonl`],
  ["find_schema", `find ${MOUNT}/public/ -name schema.json`],
  ["jq_titles", `jq '.title' ${MOUNT}/public/tables/books/rows.jsonl`],
  ["pipe_grep_c", `cat ${MOUNT}/public/tables/books/rows.jsonl | grep -c ada`],
  ["cat_view_rows", `cat ${MOUNT}/public/views/recent_books/rows.jsonl`],
  ["wc_l_view", `wc -l ${MOUNT}/public/views/recent_books/rows.jsonl`],
  ["safeguard_cat_truncates", `cat ${MOUNT}/public/tables/books/rows.jsonl`],
  ["safeguard_cat_pipe_uncapped", `cat ${MOUNT}/public/tables/books/rows.jsonl | wc -l`],
];

async function seed(client: pg.Client): Promise<void> {
  await client.query("DROP VIEW IF EXISTS recent_books");
  await client.query("DROP TABLE IF EXISTS books");
  await client.query("DROP TABLE IF EXISTS authors");
  await client.query(
    "CREATE TABLE books (id int PRIMARY KEY, title text, author text, year int, rating double precision)",
  );
  await client.query("CREATE TABLE authors (id int PRIMARY KEY, name text, books int)");
  for (const [id, title, author, year, rating] of BOOKS) {
    await client.query(
      "INSERT INTO books (id, title, author, year, rating) VALUES ($1, $2, $3, $4, $5)",
      [id, title, author, year, rating],
    );
  }
  for (const [id, name, books] of AUTHORS) {
    await client.query("INSERT INTO authors (id, name, books) VALUES ($1, $2, $3)", [
      id,
      name,
      books,
    ]);
  }
  await client.query("CREATE VIEW recent_books AS SELECT * FROM books WHERE year >= 2022");
  await client.query("ANALYZE books");
  await client.query("ANALYZE authors");
}

async function run(ws: Workspace, name: string, cmd: string): Promise<void> {
  const result = await ws.execute(cmd);
  const out = DEC.decode(result.stdout);
  process.stdout.write(`=== ${name} ===\n`);
  process.stdout.write(out.endsWith("\n") ? out : out + "\n");
  if (name.startsWith("safeguard_")) {
    const err = DEC.decode(result.stderr);
    if (err) process.stdout.write(err.endsWith("\n") ? err : err + "\n");
  }
}

function setCatSafeguard(ws: Workspace, maxLines: number): void {
  const sg = new CommandSafeguard({ maxLines });
  for (const m of ws.registry.allMounts()) m.commandSafeguards.set("cat", sg);
  if (ws.registry.defaultMount !== null) {
    ws.registry.defaultMount.commandSafeguards.set("cat", sg);
  }
}

async function main(): Promise<void> {
  const seedClient = new pg.Client({ connectionString: DSN });
  await seedClient.connect();
  try {
    await seed(seedClient);
  } finally {
    await seedClient.end();
  }
  const resource = new PostgresResource({ dsn: DSN, maxReadRows: 200 });
  const ws = new Workspace({ [MOUNT]: resource }, { mode: MountMode.READ });
  try {
    for (const [name, cmd] of CASES) {
      if (name === "safeguard_cat_truncates") setCatSafeguard(ws, 2);
      await run(ws, name, cmd);
    }
  } finally {
    await ws.close();
    await resource.close();
  }
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});

File: integ/postgres.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import os

import asyncpg
from cases import run_not_found

from mirage import MountMode, Workspace
from mirage.resource.postgres import PostgresConfig, PostgresResource
from mirage.types import CommandSafeguard

DSN = os.environ.get("POSTGRES_DSN",
                     "postgres://mirage:mirage@localhost:55432/mirage_integ")
MOUNT = "/pg"

BOOKS = [
    (1, "alpha", "ada", 2020, 4.5),
    (2, "beta", "ben", 2021, 3.2),
    (3, "gamma", "cara", 2022, 5.0),
    (4, "delta", "ada", 2023, 4.0),
    (5, "epsilon", "ben", 2024, 2.5),
]

AUTHORS = [
    (1, "ada", 2),
    (2, "ben", 2),
    (3, "cara", 1),
]

CASES: list[tuple[str, str]] = [
    ("ls_root", f"ls {MOUNT}/"),
    ("ls_schema", f"ls {MOUNT}/public/"),
    ("ls_tables", f"ls {MOUNT}/public/tables/"),
    ("ls_views", f"ls {MOUNT}/public/views/"),
    ("ls_entity", f"ls {MOUNT}/public/tables/books/"),
    ("tree", f"tree -L 3 {MOUNT}/public/"),
    ("cat_schema_json", f"cat {MOUNT}/public/tables/books/schema.json"),
    ("cat_rows", f"cat {MOUNT}/public/tables/books/rows.jsonl"),
    ("head_2", f"head -n 2 {MOUNT}/public/tables/books/rows.jsonl"),
    ("tail_2", f"tail -n 2 {MOUNT}/public/tables/books/rows.jsonl"),
    ("wc_l_books", f"wc -l {MOUNT}/public/tables/books/rows.jsonl"),
    ("wc_default_books", f"wc {MOUNT}/public/tables/books/rows.jsonl"),
    ("wc_l_authors", f"wc -l {MOUNT}/public/tables/authors/rows.jsonl"),
    ("stat_rows", f"stat {MOUNT}/public/tables/books/rows.jsonl"),
    ("grep_c_title", f"grep -c title {MOUNT}/public/tables/books/rows.jsonl"),
    ("grep_ada", f"grep ada {MOUNT}/public/tables/books/rows.jsonl"),
    ("grep_e_multi",
     f"grep -n -e ada -e ben {MOUNT}/public/tables/books/rows.jsonl"),
    ("grep_schema_scope", f"grep ada {MOUNT}/public/tables/"),
    ("rg_schema_scope", f"rg ben {MOUNT}/public/"),
    ("find_rows", f"find {MOUNT}/public/ -name rows.jsonl"),
    ("find_schema", f"find {MOUNT}/public/ -name schema.json"),
    ("jq_titles", f"jq '.title' {MOUNT}/public/tables/books/rows.jsonl"),
    ("pipe_grep_c",
     f"cat {MOUNT}/public/tables/books/rows.jsonl | grep -c ada"),
    ("cat_view_rows", f"cat {MOUNT}/public/views/recent_books/rows.jsonl"),
    ("wc_l_view", f"wc -l {MOUNT}/public/views/recent_books/rows.jsonl"),
    ("safeguard_cat_truncates", f"cat {MOUNT}/public/tables/books/rows.jsonl"),
    ("safeguard_cat_pipe_uncapped",
     f"cat {MOUNT}/public/tables/books/rows.jsonl | wc -l"),
]


async def _seed(conn: asyncpg.Connection) -> None:
    await conn.execute("DROP VIEW IF EXISTS recent_books")
    await conn.execute("DROP TABLE IF EXISTS books")
    await conn.execute("DROP TABLE IF EXISTS authors")
    await conn.execute("CREATE TABLE books (id int PRIMARY KEY, title text, "
                       "author text, year int, rating double precision)")
    await conn.execute(
        "CREATE TABLE authors (id int PRIMARY KEY, name text, books int)")
    await conn.executemany(
        "INSERT INTO books (id, title, author, year, rating) "
        "VALUES ($1, $2, $3, $4, $5)", BOOKS)
    await conn.executemany(
        "INSERT INTO authors (id, name, books) VALUES ($1, $2, $3)", AUTHORS)
    await conn.execute(
        "CREATE VIEW recent_books AS SELECT * FROM books WHERE year >= 2022")
    await conn.execute("ANALYZE books")
    await conn.execute("ANALYZE authors")


async def _run(ws: Workspace, name: str, cmd: str) -> None:
    result = await ws.execute(cmd)
    out = await result.stdout_str()
    print(f"=== {name} ===")
    print(out, end="" if out.endswith("\n") else "\n")
    if name.startswith("safeguard_"):
        err = await result.stderr_str()
        if err:
            print(err, end="" if err.endswith("\n") else "\n")


def _set_cat_safeguard(ws: Workspace, max_lines: int) -> None:
    sg = CommandSafeguard(max_lines=max_lines)
    mounts = list(ws._registry._mounts)
    if ws._registry.default_mount is not None:
        mounts.append(ws._registry.default_mount)
    for m in mounts:
        m.command_safeguards["cat"] = sg


async def main() -> None:
    conn = await asyncpg.connect(DSN)
    try:
        await _seed(conn)
    finally:
        await conn.close()
    resource = PostgresResource(
        config=PostgresConfig(dsn=DSN, max_read_rows=200))
    ws = Workspace({MOUNT: resource}, mode=MountMode.READ)
    for name, cmd in CASES:
        if name == "safeguard_cat_truncates":
            _set_cat_safeguard(ws, max_lines=2)
        await _run(ws, name, cmd)
    await run_not_found(ws, MOUNT)
    await resource.accessor.close()


if __name__ == "__main__":
    asyncio.run(main())
